cpmpy/expressions/python_builtins.py

```diff
@@ -100,3 +102,61 @@ def sum(iterable):
     if not any(isinstance(elem, Expression) for elem in iterable):
         return np.sum(iterable)
     return Operator("sum", iterable)
+
+@hook(bool, int)
+def __and__(self, other):
+    if not (isinstance(other,_IntVarImpl) or isinstance(other, _BoolVarImpl) ):
+        return orig(self, other)
+    else:
+
+        if not self.is_bool():
+            raise CPMpyException(f"Logical conjunction involving an integer ({self}) is not allowed")
+
+        if isinstance(self,int):
+            self = (self == 1)  # convert to True or False
+
+        if not isinstance(other,_BoolVarImpl):
+            raise CPMpyException(f"Logical conjunction involving an IntVar ({other}) is not allowed")
+
+        return Operator("and", [self, other])
```

Comment:
Making this code a bit shorter, we can actually do the following:

```python
@hook(bool, int)
def __and__(self, other):
    if isinstance(other, Expression):
        return other.__rand__(self)
    return orig(self, other)
```
This avoids a lot of duplication.

Reply:
I assume you mean 

`        return other.__and__(self)
`

and not 

`        return other.__rand__(self)`

Reply:
No, in Python, all `__` operators are also reversed.
As we [implemented both](https://github.com/CPMpy/cpmpy/blob/9fba76cc89845deb99f01a0b396021cc7f708924/cpmpy/expressions/core.py#L214), we can simply call the reversed operator from here and everything should work as is.

Reply:
However, we still would need to check if the self part is bool or not, and also if the other part is bool (i.e., as it is an expression in this case, it can be an IntVar, where we throw an expression). So, it should call the appropriate function (like the newly added ones in _BoolVarImpl and _IntVarImpl